### DISCUSSION TITLE
Installation on ALCF Theta

### DISCUSSION THREAD

**Discussion by smpark7** [idaholab/moose#20082] [Q&A Getting Started] [answered]
I'm currently trying to install a MOOSE App on Theta. I don't have any specific requirements concerning PETSc and libMesh versions. Can anyone provide me with the appropriate environment for installing MOOSE on Theta?

I tried the following:
```module switch PrgEnv-intel PrgEnv-gnu
module load cray-hdf5-parallel
module load cray-fftw
module load cray-petsc
module load cray-tpsl
module load cray-python
export CC=cc
export CXX=CC
export FC=ftn
export EPYTHON=python3.8
export PETSCDIR=$PETSC_DIR
export CRAYPE_LINK_TYPE=dynamic
export CRAY_ADD_RPATH=yes
```
which I (blindly) adapted from my experience with another supercomputer. This resulted in a seemingly successful installation, but when I run an input file (that works on other machines), the simulation run gets killed by the OOM killer despite using only ~32GB of the available 192GB memory per node.

**Comment by GiudGiud** [comment, marked as answer]
Hello

Does the test suite pass?
How did you measure the 32 GB ?
How did you request the 192 GB ?

Guillaume

**Reply by smpark7** [reply]
Hello,

Thanks for the quick reply. I have trouble running `run_tests` despite providing `export MOOSE_MPI_COMMAND=aprun` and `alias mpiexec=aprun`. I get the following error message for all tests:
```nts.nts: Working Directory: /gpfs/mira-home/smpark3/projects/moltres/tests/nts
nts.nts: Running command: aprun -n 1 /gpfs/mira-home/smpark3/projects/moltres/moltres-opt -i nts.i --no-gdb-backtrace
nts.nts: apsched: request exceeds max nodes, alloc
nts.nts: apsched: request exceeds max nodes, alloc
nts.nts: 
nts.nts: 
nts.nts: Exit Code: 1
nts.nts: ################################################################################
nts.nts: Tester failed, reason: CRASH
nts.nts: 
nts.nts ....................................................................................... FAILED (CRASH)
```
The test input files seem to run successfully when I manually run `aprun -n 64 ...` individually, but this approach obviously does not compare the output against the gold copy. I measured 32 GB using MemoryUsage postprocessor with the default settings which report total memory usage across all ranks. I assumed 192 GB of memory is provided based on Theta's [machine info](https://www.alcf.anl.gov/support-center/theta/theta-thetagpu-overview).

Sun Myung

**Reply by GiudGiud** [reply]
Well it depends on the cluster. I think if you requested a full node, you should get the whole memory for it. Is that what you are doing with -n 1 ?

"apsched: request exceeds max nodes, alloc" is this the oom output?
aprun -n 1 on the inputs individually works too?
Doesnt theta have some minimum ressource request? Like at least 4 nodes?

**Reply by smpark7** [reply]
No, the OOM killer error occurred when I ran a non-test input file with `aprun -n 64` on 1 compute node, whereas the `apsched: request exceeds max nodes, alloc` error occurred when I tried to run `run_tests`.

Yes `aprun -n 1` works when individually running tests. I ran it on the debug queue which allows me to request 1 to 8 compute nodes.

**Reply by GiudGiud** [reply]
Are you using `run_tests` on the debug queue as well?
This troubleshooting page: http://www.archer.ac.uk/documentation/troubleshooting.php (for another cluster, is there a theta one?)
is saying that aprun is being run on the head node instead of within a job script.

I m not familiar with aprun. Is this more of an mpirun command or more of an interactive qsub command

**Reply by smpark7** [reply]
Yes I ran `run_tests` on the debug queue as well. Unfortunately, Theta's user guide doesn't seem to have troubleshooting information relevant to my issues.

aprun seems to function similarly to mpirun; I just replace mpirun with aprun in my commands. aprun is responsible for starting jobs on the compute nodes from the service nodes from which we send commands. Theta provides this illustration as an overview of what qsub and aprun do:

![image](https://user-images.githubusercontent.com/30237775/150205916-0a94dd36-dcc7-4c54-87cc-5f1ab934ee0a.png)

**Reply by GiudGiud** [reply]
ah ok so you're running all this from a qsub interactive session?
why would you get this `apsched: request exceeds max nodes, alloc` then though...
Seems like 1 should be ok.

**Reply by smpark7** [reply]
Yes it's very perplexing to get this error even though I'm in the interactive session and `run_tests` ran `aprun -n 1`.

Regarding the OOM issue, I tried running the same input file with fewer MPI ranks per node (32 vs 64 previously). I still got the same OOM killer triggered at the same point during the simulation (right after 5 time steps). This seems to imply that the OOM issue is not hardware-related.

I found an old [thread](https://groups.google.com/g/moose-users/c/fOkN6qQSXGM/m/8H-qqXelCAAJ) from 2018 in the moose-users group about installing MOOSE on Theta. Could the comment from Fande about static vs dynamic libs be relevant to my OOM killer issue?

**Reply by GiudGiud** [reply]
Oh great let's tag @fdkong since he was running stuff on theta.
I dont think it's related though. I'd double check that your interactive session has requested enough memory. There should be a way to get that from `qstat`

**Reply by fdkong** [reply]
Does "`aprun -n 1 /gpfs/mira-home/smpark3/projects/moltres/moltres-opt -i nts.i`" run if you type it manually? 

I did not see any indication that this is related to how did you build moose. I ran MOOSE a couple of years ago but I did not do "run_tests".   I tried to use a job script to launch the simulation I want.

**Reply by smpark7** [reply]
Yes, all test input files run if I run them individually like that. From manually scanning through some of the postprocessor values, the values seem to indicate that the results match the gold copies. So I'm not too concerned about `run_tests` not running correctly.

Do you have any advice/ideas about the OOM killer issue? I tried asking the ALCF support desk, but they recommended directing my question to the MOOSE developers.

**Reply by fdkong** [reply]
Regarding OOM, could you let me know what is the total DoFs of the simulation? What are the solver options that you are using? It could be a true OOM.  How many nodes are you using?

**Reply by smpark7** [reply]
The simulation is a MultiApp run. The main app is a Transient simulation with 160k DOFs and 200k aux DOFs, NEWTON solve, asm with lu (1 overlap), while the sub app is a InversePowerMethod simulation with 560k DOFs and 240 aux DOFs, NEWTON solve, asm with lu (1 overlap).

I ran it on 8 compute nodes twice. When run on 64 ranks per node (512 total ranks), the latest logged memory use (according to MemoryUsage) before OOM is ~32 GB total. When run on 32 ranks per node (256 total ranks), the latest logged memory use before OOM is ~10 GB total. Both runs encounter the OOM killer during the fifth time step, in between the sub app and main app runs.

It's weird because I successfully ran this input file in early 2021 on Theta under a different project allocation. I did a clean install of Moltres using the same environment, but now I encounter this OOM killer issue.

**Reply by fdkong** [reply]
This size of problem should be fine on eight compute nodes.  IIRC, there is 192 GB per node.  You could compile the code to a "devel" mode to trigger more libmesh checks.  It could be some error during the simulation, and this kind of error we usually do not check in the opt mode because of the performance.  If this type of error happens in the simulation, we will receive a segfault or other error messages.

Even though it is unlikely to be a real out of memory, it does not hurt to try "asm with ilu".  LU is a lot more expensive than ILU in memory

**Reply by smpark7** [reply]
Thanks for the suggestions. I'll try out devel mode and see what I get. ILU comes with a slight hit on performance for my problem, but I agree it is an option.

**Reply by smpark7** [reply]
Somehow the OOM error doesn't occur if I use `aprun -n 256 -N 64 -d 1 -j 1`, which is baffling to me because `-d` and `-j` should both default to 1 if they're not specified.
